Modules/posixmodule.c

```diff
@@ -4229,7 +4229,224 @@ os_listdir_impl(PyObject *module, path_t *path)
 #endif
 }
 
+
 #ifdef MS_WINDOWS
+
+/*[clinic input]
+os.listdrives
+
+Return a list containing the names of drives in the system.
+
+A drive name typically looks like 'C:\\'.
+
+[clinic start generated code]*/
+
+static PyObject *
+os_listdrives_impl(PyObject *module)
+/*[clinic end generated code: output=aaece9dacdf682b5 input=1af9ccc9e583798e]*/
+{
+    wchar_t defaultBuffer[64];
```

Comment:
```suggestion
    wchar_t defaultBuffer[129];
```

`GetLogicalDriveStringsW()` simply bitshifts the result from `GetLogicalDrives()`, storing a drive string for each defined drive. The latter returns a `DWORD`, but only 26 drives can be defined instead of 32. That's never going to change, but the size is small enough that we may as well use a default buffer that can hold 32 drives -- 129 wide characters. A bigger buffer will never be required. This function can then be simplified a lot, by not requiring any heap allocation.

Reply:
Here's the [example](https://learn.microsoft.com/en-us/windows/win32/memory/obtaining-a-file-name-from-a-file-handle) from the docs, which calls `GetLogicalDriveStrings()` with a significantly over-allocated buffer length of 512 (`BUFSIZE`) characters. It only cares about the return value in terms of success (non-zero) and failure (0).

Reply:
I did wonder whether it had a fixed maximum size. I guess since it's the old API it's unlikely to grow, and also we're unlikely to ever see a device with that many drives we'll probably never find out. I'll go to the simpler code.

Reply:
> we're unlikely to ever see a device with that many drives we'll probably never find out.

It's easy enough to subst every possible drive. It's an old way to get around the `MAX_PATH` limit since the target path set by `DefineDosDeviceW()` supports up to about 4K characters. But then `GetFinalPathNameByHandleW()` spoils the fun by resolving the substitute drive back to the canonical volume path, as it should do.

Reply:
We'd have to know every possible drive to subst them all, which means we already know the maximum.

I was more thinking of a hypothetical future where a future Windows suddenly introduces a new set of drive letters and we fail to notice. But that seems very unlikely

Reply:
> I was more thinking of a hypothetical future where a future Windows suddenly introduces a new set of drive letters and we fail to notice. But that seems very unlikely

The defined logical drives are tracked in a 32-bit bitmap in the kernel, and the `GetLogicalDrives()` API function is based on this. Also, the object manager only reserves devices named "[A-Za-z]:" when updating the drive bitmap of a logon session. In principle, this could all be extended to support more characters, but they'd need to add new API functions such as `GetLogicalDrivesEx()`. This would be a compatibility nightmare.

The classic "[A-Z]:" logical drive names are here to stay, but systems with many volumes that need persistent names should use filesystem mountpoints instead of drive-letter names. For me, it's easier to remember or identify a descriptive path than having to remember what's on drive "Q:".